File: jeph2sworm-core/jeph2sworm/llm/cohere_provider.py
"""Cohere Provider - Command R+, Command R."""

from __future__ import annotations

from typing import List

import structlog

from jeph2sworm.llm.base_provider import BaseProvider, LLMMessage, LLMResponse

logger = structlog.get_logger()

MODELS = [
    "command-r-plus",
    "command-r",
    "command-light",
]

TASK_BEST = {
    "coding": "command-r-plus",
    "planning": "command-r-plus",
    "reasoning": "command-r-plus",
    "design": "command-r",
    "testing": "command-light",
    "general": "command-r",
}


class CohereProvider(BaseProvider):
    provider_name = "cohere"
    supports_function_calling = True
    supports_vision = False
    max_context_window = 128_000

    def available_models(self) -> List[str]:
        return MODELS

    def best_model_for(self, task: str) -> str:
        return TASK_BEST.get(task, "command-r")

    async def complete(
        self,
        messages: List[LLMMessage],
        model: str,
        temperature: float = 0.7,
        max_tokens: int = 4096,
        **kwargs,
    ) -> LLMResponse:
        import litellm

        response = await litellm.acompletion(
            model=f"cohere/{model}" if not model.startswith("cohere/") else model,
            messages=[m.model_dump() for m in messages],
            temperature=temperature,
            max_tokens=max_tokens,
            api_key=self.api_key,
            **kwargs,
        )

        choice = response.choices[0]
        usage = dict(response.usage) if response.usage else {}

        return LLMResponse(
            content=choice.message.content or "",
            model=model,
            provider=self.provider_name,
            usage=usage,
            finish_reason=choice.finish_reason or "stop",
        )

File: jeph2sworm-vscode/src/providers/cohere-provider.ts
/**
 * Cohere Provider — proxy to Cohere Command-R / Command-R+ models.
 */

import { BaseProvider, LLMMessage, LLMResponse, ProviderConfig } from './base-provider';

export class CohereProvider extends BaseProvider {
  readonly name = 'cohere';
  readonly models = ['command-r-plus', 'command-r', 'command-light'];

  constructor(config: ProviderConfig) {
    super({ baseUrl: 'https://api.cohere.ai/v2', ...config });
  }

  async complete(messages: LLMMessage[], model?: string): Promise<LLMResponse> {
    const chosen = model ?? this.config.defaultModel ?? 'command-r-plus';
    const res = await fetch(`${this.config.baseUrl}/chat`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.config.apiKey}`,
      },
      body: JSON.stringify({ model: chosen, messages }),
    });
    const data = await res.json() as any;
    return {
      content: data.message?.content?.[0]?.text ?? '',
      model: chosen,
      tokensUsed:
        (data.usage?.tokens?.input_tokens ?? 0) + (data.usage?.tokens?.output_tokens ?? 0),
      finishReason: data.finish_reason ?? 'unknown',
    };
  }

  bestModelFor(task: 'coding' | 'planning' | 'design' | 'quick'): string {
    switch (task) {
      case 'coding':
        return 'command-r-plus';
      case 'planning':
        return 'command-r-plus';
      case 'design':
        return 'command-r';
      case 'quick':
        return 'command-light';
    }
  }
}
